source/slang/slang-check-overload.cpp

```diff
@@ -1557,10 +1582,32 @@ namespace Slang
             if (externExportDiff)
                 return externExportDiff;
 
+            // We need to consider the distance of the declarations to the global scope to resolve this case:
+            //      float f(float x);
+            //      struct S
+            //      {
+            //          float f(float x);
+            //          float g(float y) { return f(y); }   // will call S::f() instead of ::f()
+            //      }
+            //  we will count the distance from the reference site to the declaration in the scope tree.
+            //  NOTE: DON'T do this for the generic function, because for the generic function compare, we only check the
+            //  generic parameter/arguments, not the function parameters/arguments. This could break the assumption
+            //  for this logic where we assume that all the candidates are valid.
+            //  This logic will be evaluated later when "Flavor" becomes "Func" or "Expr", because by then the type checks for
+            //  the candidate will be done.
+            if ((left->flavor != OverloadCandidate::Flavor::Generic && left->flavor != OverloadCandidate::Flavor::UnspecializedGeneric) ||
```

Comment:
I feel like this needs to be && instead of ||?
Can you change it to use early return instead of nested if?
I feel like the overloadRankDiff logic down below should be guarded by the flavor check too.

Reply:
Why does it needs to by &&?
Either left or right is generic, we can't make the compare here anyway.

> I feel like the overloadRankDiff logic down below should be guarded by the flavor check too

I'm not very familiar with this, just did a search, it's a attribute `[OverloadRank(#)]`, if this can be applied to generic too, then yes, we should guard this. Because only the valid candidates should be in the rank compare, correctness is always the highest priority.

I will make a change to include this and add some commend.

Reply:
push a new change.